fedora_messaging/twisted/protocol.py

```diff
@@ -0,0 +1,296 @@
+            result = yield self._channel.queue_declare(
+                queue=binding['queue_name'],
+                durable=True,
+                auto_delete=binding.get("queue_auto_delete", False),
+                arguments=binding.get('queue_arguments'),
+            )
+            queue_name = result.method.queue
+            yield self._channel.queue_bind(
+                queue=queue_name,
+                exchange=binding['exchange'],
+                routing_key=binding['routing_key'],
+            )
+            self._queues.add(queue_name)
+        log.msg("AMQP bindings declared", system=self.name,
+                logLevel=logging.DEBUG)
+
+    @defer.inlineCallbacks
+    def _read(self, queue_object):
+        while self._running:
+            try:
+                channel, delivery_frame, properties, body = \
+                    yield queue_object.get()
+            except (
+                    error.ConnectionDone, pika.exceptions.ChannelClosedByClient
+                    ):
+                # This is deliberate.
+                log.msg("Closing the read loop on the producer.",
+                        system=self.name, logLevel=logging.DEBUG)
+                break
+            except pika.exceptions.ChannelClosed as e:
+                log.err(e, system=self.name)
+                break
+            except pika.exceptions.ConsumerCancelled:
+                log.msg("Consumer cancelled, quitting the read loop.",
+                        system=self.name)
+                break
+            except Exception as e:
+                log.err("Failed getting the next message in the queue, "
+                        "stopping.", system=self.name)
+                log.err(e, system=self.name)
+                break
+            if body:
+                yield self._on_message(delivery_frame, properties, body)
+
+    @defer.inlineCallbacks
+    def _on_message(self, delivery_frame, properties, body):
+        """
+        Callback when a message is received from the server.
+
+        This method wraps a user-registered callback for message delivery. It
+        decodes the message body, determines the message schema to validate the
+        message with, and validates the message before passing it on to the
+        user callback.
+
+        This also handles acking, nacking, and rejecting messages based on
+        exceptions raised by the consumer callback. For detailed documentation
+        on the user-provided callback, see the user guide on consuming.
+
+        Args:
+            delivery_frame (pika.spec.Deliver): The delivery frame which
+                includes details about the message like content encoding and
+                its delivery tag.
+            properties (pika.spec.BasicProperties): The message properties like
+                the message headers.
+            body (bytes): The message payload.
+            message_callback (callable): the function that will be called with
```

Comment:
This looks to no longer be a method argument and is instead set in the `setupRead` call.